langchain/src/document_loaders/web/github.ts

```diff
@@ -141,34 +147,82 @@ export class GithubRepoLoader
     );
   }
 
-  private async processDirectory(
-    path: string,
-    documents: Document[]
-  ): Promise<void> {
-    try {
-      const files = await this.fetchRepoFiles(path);
+  /**
+   * Takes the file info and wrap it in a promise that will resolve to the file content and metadata
+   * @param file
+   * @returns
+   */
+  private async fetchFileContentWrapper(
+    file: GithubFile
+  ): Promise<GetContentResponse> {
+    const fileContent = await this.fetchFileContent(file).catch((error) => {
+      this.handleError(`Failed wrap file content: ${file}, ${error}`);
+    });
+    return {
+      contents: fileContent || "",
+      metadata: { source: file.path },
+    };
+  }
+
+  /**
+   * Maps a list of files / directories to a list of promises that will fetch the file / directory contents
+   */
+  private async getCurrDirFilesPromises(
+    files: GithubFile[]
+  ): Promise<Promise<GetContentResponse>[]> {
+    const currDirFilePromises: Promise<GetContentResponse>[] = [];
+    // Directories have nested files / directories, which is why this is a list of promises of promises
+    const currDirDirPromises: Promise<Promise<GetContentResponse>[]>[] = [];
 
-      for (const file of files) {
-        if (!(await this.shouldIgnore(file.path, file.type))) {
-          if (file.type !== "dir") {
-            try {
-              const fileContent = await this.fetchFileContent(file);
-              const metadata = { source: file.path };
-              documents.push(
-                new Document({ pageContent: fileContent, metadata })
-              );
-            } catch (e) {
-              this.handleError(
-                `Failed to fetch file content: ${file.path}, ${e}`
-              );
-            }
-          } else if (this.recursive) {
-            await this.processDirectory(file.path, documents);
+    for (const file of files) {
+      if (!this.shouldIgnore(file.path, file.type)) {
+        if (file.type !== "dir") {
+          try {
+            currDirFilePromises.push(this.fetchFileContentWrapper(file));
+          } catch (e) {
+            this.handleError(
+              `Failed to fetch file content: ${file.path}, ${e}`
+            );
           }
+        } else if (this.recursive) {
+          currDirDirPromises.push(this.processDirectory(file.path));
         }
       }
+    }
+
+    const curDirDirectories: Promise<GetContentResponse>[][] =
+      await Promise.all(currDirDirPromises);
```

Comment:
Can we set some kind of max concurrency here?

Something like:

https://github.com/hwchase17/langchainjs/blob/main/langchain/src/schema/runnable.ts#L60